// forta-agent/dist/sdk/finding.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Finding = exports.FindingType = exports.FindingSeverity = void 0;
var label_1 = require("./label");
var utils_1 = require("./utils");
var FindingSeverity;
(function (FindingSeverity) {
    FindingSeverity[FindingSeverity["Unknown"] = 0] = "Unknown";
    FindingSeverity[FindingSeverity["Info"] = 1] = "Info";
    FindingSeverity[FindingSeverity["Low"] = 2] = "Low";
    FindingSeverity[FindingSeverity["Medium"] = 3] = "Medium";
    FindingSeverity[FindingSeverity["High"] = 4] = "High";
    FindingSeverity[FindingSeverity["Critical"] = 5] = "Critical";
})(FindingSeverity = exports.FindingSeverity || (exports.FindingSeverity = {}));
var FindingType;
(function (FindingType) {
    FindingType[FindingType["Unknown"] = 0] = "Unknown";
    FindingType[FindingType["Exploit"] = 1] = "Exploit";
    FindingType[FindingType["Suspicious"] = 2] = "Suspicious";
    FindingType[FindingType["Degraded"] = 3] = "Degraded";
    FindingType[FindingType["Info"] = 4] = "Info";
    FindingType[FindingType["Scam"] = 5] = "Scam";
})(FindingType = exports.FindingType || (exports.FindingType = {}));
var Finding = (function () {
    function Finding(name, description, alertId, protocol, severity, type, metadata, addresses, labels, uniqueKey, source, timestamp) {
        this.name = name;
        this.description = description;
        this.alertId = alertId;
        this.protocol = protocol;
        this.severity = severity;
        this.type = type;
        this.metadata = metadata;
        this.addresses = addresses;
        this.labels = labels;
        this.uniqueKey = uniqueKey;
        this.source = source;
        this.timestamp = timestamp;
    }
    Finding.prototype.toString = function () {
        return JSON.stringify(__assign(__assign({}, this), { severity: FindingSeverity[this.severity], type: FindingType[this.type], labels: this.labels.map(function (l) { return Object.assign(l, {
                entityType: label_1.EntityType[l.entityType]
            }); }) }), null, 2);
    };
    Finding.from = function (findingInput) {
        return this.fromObject(findingInput);
    };
    Finding.fromObject = function (_a) {
        var name = _a.name, description = _a.description, alertId = _a.alertId, _b = _a.protocol, protocol = _b === void 0 ? 'ethereum' : _b, severity = _a.severity, type = _a.type, _c = _a.timestamp, timestamp = _c === void 0 ? new Date() : _c, _d = _a.metadata, metadata = _d === void 0 ? {} : _d, _e = _a.addresses, addresses = _e === void 0 ? [] : _e, _f = _a.labels, labels = _f === void 0 ? [] : _f, _g = _a.uniqueKey, uniqueKey = _g === void 0 ? '' : _g, _h = _a.source, source = _h === void 0 ? {} : _h;
        (0, utils_1.assertIsNonEmptyString)(name, 'name');
        (0, utils_1.assertIsNonEmptyString)(description, 'description');
        (0, utils_1.assertIsNonEmptyString)(alertId, 'alertId');
        (0, utils_1.assertIsNonEmptyString)(protocol, 'protocol');
        (0, utils_1.assertIsFromEnum)(severity, FindingSeverity, 'severity');
        (0, utils_1.assertIsFromEnum)(type, FindingType, 'type');
        labels = labels.map(function (l) { return l instanceof label_1.Label ? l : label_1.Label.fromObject(l); });
        return new Finding(name, description, alertId, protocol, severity, type, metadata, addresses, labels, uniqueKey, source, timestamp);
    };
    return Finding;
}());
exports.Finding = Finding;
