// forta-agent/dist/cli/utils/get.subscription.alerts.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.provideGetSubscriptionAlerts = exports.TEN_MINUTES_IN_MS = void 0;
var _1 = require(".");
exports.TEN_MINUTES_IN_MS = 600000;
function provideGetSubscriptionAlerts(getAlerts) {
    (0, _1.assertExists)(getAlerts, "getAlerts");
    var seenAlerts = new Map();
    return function getSubscriptionAlerts(subscriptions) {
        return __awaiter(this, void 0, void 0, function () {
            var queries, _i, subscriptions_1, subscription, alertArrays, alerts, _a, alertArrays_1, alertArray, _b, alertArray_1, alert_1;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        if (subscriptions.length == 0)
                            return [2, []];
                        queries = [];
                        for (_i = 0, subscriptions_1 = subscriptions; _i < subscriptions_1.length; _i++) {
                            subscription = subscriptions_1[_i];
                            queries.push(runQuery(subscription, getAlerts));
                        }
                        return [4, Promise.all(queries)];
                    case 1:
                        alertArrays = _c.sent();
                        alerts = [];
                        for (_a = 0, alertArrays_1 = alertArrays; _a < alertArrays_1.length; _a++) {
                            alertArray = alertArrays_1[_a];
                            for (_b = 0, alertArray_1 = alertArray; _b < alertArray_1.length; _b++) {
                                alert_1 = alertArray_1[_b];
                                if (seenAlerts.has(alert_1.hash))
                                    continue;
                                alerts.push(alert_1);
                                seenAlerts.set(alert_1.hash, true);
                            }
                        }
                        return [2, alerts];
                }
            });
        });
    };
}
exports.provideGetSubscriptionAlerts = provideGetSubscriptionAlerts;
function runQuery(subscription, getAlerts) {
    var _a, _b, _c;
    return __awaiter(this, void 0, void 0, function () {
        var alerts, query, response, pageSize, shouldRetryFromError, chainId, botId, alertId, alertIds, e_1;
        var _d;
        return __generator(this, function (_e) {
            switch (_e.label) {
                case 0:
                    alerts = [];
                    pageSize = 1000;
                    shouldRetryFromError = false;
                    _e.label = 1;
                case 1:
                    _e.trys.push([1, 3, , 4]);
                    chainId = subscription.chainId, botId = subscription.botId, alertId = subscription.alertId, alertIds = subscription.alertIds;
                    query = {
                        botIds: [botId],
                        createdSince: exports.TEN_MINUTES_IN_MS,
                        first: pageSize,
                        startingCursor: response === null || response === void 0 ? void 0 : response.pageInfo.endCursor,
                    };
                    if (chainId) {
                        query.chainId = chainId;
                    }
                    if (alertId) {
                        query.alertIds = [alertId];
                    }
                    if (alertIds === null || alertIds === void 0 ? void 0 : alertIds.length) {
                        if ((_a = query.alertIds) === null || _a === void 0 ? void 0 : _a.length) {
                            (_d = query.alertIds).push.apply(_d, alertIds);
                        }
                        else {
                            query.alertIds = alertIds;
                        }
                    }
                    return [4, getAlerts(query)];
                case 2:
                    response = _e.sent();
                    shouldRetryFromError = false;
                    alerts.push.apply(alerts, response.alerts);
                    return [3, 4];
                case 3:
                    e_1 = _e.sent();
                    if (((_b = e_1.response) === null || _b === void 0 ? void 0 : _b.status) === 500) {
                        pageSize = Math.floor(pageSize / 2);
                        shouldRetryFromError = pageSize > 1;
                    }
                    else {
                        throw e_1;
                    }
                    return [3, 4];
                case 4:
                    if (shouldRetryFromError || ((_c = response === null || response === void 0 ? void 0 : response.pageInfo) === null || _c === void 0 ? void 0 : _c.hasNextPage)) return [3, 1];
                    _e.label = 5;
                case 5: return [2, alerts];
            }
        });
    });
}
